java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java

```diff
@@ -0,0 +1,136 @@
+import org.sonar.plugins.java.api.tree.SwitchStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+@Rule(
+  key = "S2259",
+  name = "Null pointers should not be dereferenced",
+  priority = Priority.BLOCKER)
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("10min")
+public class NullDereferenceCheck extends SECheck implements JavaFileScanner {
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    Multimap<Tree, String> issues = ((DefaultJavaFileScannerContext) context).getSEIssues(NullDereferenceCheck.class);
+    for (Map.Entry<Tree, String> issue : issues.entries()) {
+      context.reportIssue(this, issue.getKey(), issue.getValue());
+    }
+  }
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    SymbolicValue currentVal = context.getState().peekValue();
+    if (currentVal == null) {
+      // stack is empty, nothing to do.
+      return context.getState();
+    }
+    if (syntaxNode.is(Tree.Kind.MEMBER_SELECT)) {
+      if("class".equals(((MemberSelectExpressionTree) syntaxNode).identifier().name())) {
+        // expression ClassName.class won't raise NPE.
+        return context.getState();
+      }
+
+      if (context.isNull(currentVal)) {
+        context.reportIssue(syntaxNode, this, "NullPointerException might be thrown as '" + getName(syntaxNode) + "' is nullable here");
+        return null;
+      }
+      // we dereferenced the symbolic value so we can assume it is not null
+      return currentVal.setSingleConstraint(context.getState(), ConstraintManager.NullConstraint.NOT_NULL);
+    }
+    return context.getState();
+  }
+
+  @Override
+  public void checkPostStatement(CheckerContext context, Tree syntaxNode) {
+    if (context.isNull(context.getState().peekValue()) && syntaxNode.is(Tree.Kind.SWITCH_STATEMENT)) {
+      context.reportIssue(syntaxNode, this, "NullPointerException might be thrown as '" + getName(syntaxNode) + "' is nullable here");
+      context.createSink();
+      return;
+    }
+    List<ProgramState> programStates = setNullConstraint(context, syntaxNode);
+    for (ProgramState programState : programStates) {
+      context.addTransition(programState);
+    }
+  }
+
+  private List<ProgramState> setNullConstraint(CheckerContext context, Tree syntaxNode) {
+    SymbolicValue val = context.getState().peekValue();
+    if (syntaxNode.is(Tree.Kind.NULL_LITERAL)) {
+      //invariant to check that value was correctly evaluated.
+      assert val != null && val.equals(SymbolicValue.NULL_LITERAL);
+    } else if (syntaxNode.is(Tree.Kind.METHOD_INVOCATION)) {
+      if (isAnnotatedCheckForNull((MethodInvocationTree) syntaxNode)) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Merge this if statement with the enclosing one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1066)